File: AVFoudationDemo/AVFoudationDemo/ViewController.m
//
//  ViewController.m
//  AVFoudationDemo
//
//  Created by WangWei on 17/2/17.
//  Copyright © 2017年 WangWei. All rights reserved.
//

#import "ViewController.h"
#import <AVFoundation/AVFoundation.h>
@interface ViewController ()
@property (weak, nonatomic) IBOutlet UIImageView *thumbnail;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    AVURLAsset *asset = [[AVURLAsset alloc] initWithURL:[NSURL URLWithString:@"http://fs.refanqie.com/Team.mp4"] options:nil];
    
    NSArray *keys = @[@"duration"];
    
    NSError *error = nil;
    
    AVKeyValueStatus traceStatus = [asset statusOfValueForKey:@"duration" error:&error];
    NSLog(@"%ld", traceStatus);
    
    
    [asset loadValuesAsynchronouslyForKeys:keys completionHandler:^{
        
        NSError *error = nil;
        NSLog(@"%lld", asset.duration.value / asset.duration.timescale);
        AVKeyValueStatus traceStatus = [asset statusOfValueForKey:@"duration" error:&error];
        NSLog(@"%ld", traceStatus);
        
    }];
    
    UIImage *image =  [self getThumbnail:asset];
    self.thumbnail.image = image;
    


    
}

- (UIImage *)getThumbnail:(AVURLAsset *)asset {
  
    if ([asset tracksWithMediaType:AVMediaTypeVideo].count == 0) {
        return  nil;
    }
    
    AVAssetImageGenerator *imageGenerator = [[AVAssetImageGenerator alloc] initWithAsset:asset];
    Float64 durationSecond = CMTimeGetSeconds([asset duration]);
    CMTime midpoint = CMTimeMakeWithSeconds(durationSecond / 2, 600);
    NSError *error;
    CMTime actualTime;
    
    CGImageRef halfWayImage = [imageGenerator copyCGImageAtTime:midpoint actualTime:&actualTime error:&error];
    NSString *actualTimeString = (NSString *)CFBridgingRelease(CMTimeCopyDescription(NULL, actualTime));
    NSString *requestedTimeString = (NSString *)CFBridgingRelease(CMTimeCopyDescription(NULL, midpoint));
    NSLog(@"Got halfWayImage: Asked for %@, got %@", requestedTimeString, actualTimeString);
    
    UIImage *image = [UIImage imageWithCGImage:halfWayImage];
    // Do something interesting with the image.
    CGImageRelease(halfWayImage);
    
    return image;
    
    
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: NSOperationDemo/NSOperationDemo/ViewController.swift
//
//  ViewController.swift
//  NSOperationDemo
//
//  Created by WangWei on 20/12/16.
//  Copyright © 2016年 WangWei. All rights reserved.
//

import UIKit


class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        let operation = ConOperation()
//
//        let queue = OperationQueue()
//        
//        queue.addOperation(operation)
        
        operation.start()        
        self.addObserver(self, forKeyPath: "isExecuting", options: NSKeyValueObservingOptions.new, context: nil)
        
    }
    
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        
        
        if keyPath == "isExecuting" {
            print(change?[NSKeyValueChangeKey.newKey] ?? 0)
            
        }
        
        print(change as Any)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}

